project/profiles/capacity4more/modules/c4m/user/c4m_user_profile/c4m_user_profile.module

```diff
@@ -844,3 +853,173 @@ function c4m_user_profile_init() {
     $plugin->execute($user);
   }
 }
+
+/**
+ * Render a hash for a specific user.
+ *
+ * @param mixed $account
+ *   A user account.
+ *
+ * @return string
+ *   The hash string.
+ */
+function c4m_user_profile_render_user_unsubscribe_hash($account = NULL) {
+  if (empty($account)) {
+    global $user;
+    $account = $user;
+  }
+
+  $full_name = c4m_user_name($account);
+  $wrapper = entity_metadata_wrapper('user', $account);
+  $username = $wrapper->name->value();
+  $mail = $wrapper->mail->value();
+  $organisation = $wrapper->c4m_organisation->value();
+
+  return md5($username . $mail . $full_name . $organisation . REQUEST_TIME);
+}
+
+/**
+ * Implements hook_user_presave().
+ *
+ * Adding the unsubscribe hash to the user profile when it is not a new profile.
+ */
+function c4m_user_profile_user_presave(&$edit, $account, $category) {
+  if (!$account->is_new) {
+    return;
+  }
+
+  $wrapper = entity_metadata_wrapper('user', $account);
+  $wrapper->c4m_unsubscribe_hash->set(c4m_user_profile_render_user_unsubscribe_hash($account));
+}
+
+/**
+ * Access callback; Unsubscribe a user.
+ *
+ * @param object $account
+ *   The user ID.
+ * @param string $entity_type
+ *   The flag suffix to unsubscribe from.
+ * @param string $entity_id
+ *   The ID of the entity to unsubscribe from.
+ * @param string $hash
+ *   The unsubscribe hash string.
+ *
+ * @return bool
+ *   TRUE if user has access.
+ */
+function c4m_user_profile_unsubscribe_user_access($account, $entity_type, $entity_id, $hash) {
+  if (!$account || c4m_user_is_spammer($account) || $account->status === 0) {
+    return FALSE;
+  }
+
+  $wrapper = entity_metadata_wrapper('user', $account);
+  if ($wrapper->c4m_unsubscribe_hash->value() != $hash) {
+    return FALSE;
+  }
+
+  // Validate we have an optout field for the current entity type.
+  if (!isset($wrapper->{"c4m_optout_{$entity_type}"})) {
+    return FALSE;
+  }
+
+  // Validate entity ID.
+  try {
+    $wrapper = entity_metadata_wrapper($entity_type, $entity_id);
```

Comment:
got it, good catch...

feel free to add this:
```php
// Since group is being handled by it's own flag we pass this value as the entity_type, hence for..... (keep going)
if ($entity_type == 'group') {
  $entity_type == 'node');
}
```
and add other logic that is necessary ...